ctru-rs/src/services/ir_user.rs

```diff
@@ -0,0 +1,491 @@
+                let packet_data = |idx| -> u8 {
+                    let data_buffer_offset = offset_to_data_buffer + idx;
+                    shared_mem[header_size + data_buffer_offset % data_buffer_size]
+                };
+
+                // Find out how long the payload is (payload length is variable-length encoded)
+                let (payload_length, payload_offset) = if packet_data(2) & 0x40 != 0 {
+                    // Big payload
+                    (
+                        ((packet_data(2) as usize & 0x3F) << 8) + packet_data(3) as usize,
+                        4,
+                    )
+                } else {
+                    // Small payload
+                    ((packet_data(2) & 0x3F) as usize, 3)
+                };
+
+                // Check our payload length math against what the packet info contains
+                assert_eq!(data_length, payload_offset + payload_length + 1);
+
+                // IR packets start with a magic number, so double check it
+                let magic_number = packet_data(0);
+                assert_eq!(magic_number, 0xA5);
+
+                IrUserPacket {
+                    magic_number: packet_data(0),
+                    destination_network_id: packet_data(1),
+                    payload_length,
+                    payload: (payload_offset..payload_offset + payload_length)
+                        .map(packet_data)
+                        .collect(),
+                    checksum: packet_data(payload_offset + payload_length),
+                }
+            })
+            .collect()
+    }
+
+    /// Internal helper for calling ir:USER service methods.
+    fn send_service_request(
+        &self,
+        mut request: Vec<u32>,
+        expected_response_len: usize,
+    ) -> crate::Result<Vec<u32>> {
+        let mut shared_mem_guard = IR_USER_STATE.lock().unwrap();
+        let shared_mem = shared_mem_guard.as_mut().unwrap();
+
+        // Set up the request
+        let cmd_buffer = unsafe {
+            &mut *(slice_from_raw_parts_mut(
+                ctru_sys::getThreadCommandBuffer(),
+                max(request.len(), expected_response_len),
+            ))
+        };
+        cmd_buffer[0..request.len()].copy_from_slice(&request);
+
+        // Send the request
+        unsafe {
+            ResultCode(ctru_sys::svcSendSyncRequest(shared_mem.service_handle))?;
+        }
+
+        // Handle the result returned by the service
+        ResultCode(cmd_buffer[1] as ctru_sys::Result)?;
+
+        // Copy back the response
+        request.clear();
+        request.extend_from_slice(&cmd_buffer[0..expected_response_len]);
+
+        Ok(request)
+    }
+}
+
+// Internal helper for rounding up a value to a multiple of another value.
+fn round_up(value: usize, multiple: usize) -> usize {
+    if value % multiple != 0 {
+        (value / multiple) * multiple + multiple
+    } else {
+        (value / multiple) * multiple
+    }
+}
```

Comment:
While working on the `libctru` thread-local PR I learned a neat bit twiddling trick for aligning: https://github.com/devkitPro/libctru/pull/504/files#diff-06cc21a44744e726cdc344183858d0f060e0fc6ff4355d1a1290a068c6d9c6c9R58

This version is probably more readable but maybe slightly less efficient since it has a branch in it (depending how good the optimizer is)

Reply:
I'll leave it as-is since this is only used once during initialization, and readability is more important here than performance (the optimizer may even compile it down to the optimized version).